### ISSUE TITLE
Registry: Hard error processing record

### ISSUE THREAD

**Issue by TeknoVenus** [EricZimmerman/Issues#146]
When parsing some registry hives with Registry, I see the following error logged:

```
Hard error processing record with cell signature lk at Absolute Offset: 0x1392490 with raw data: D8-FF-FF-FF-6C-6B-40-16-F2-88-84-09-C3-8A-8D-13-CC-EB-51-3C-3C-49-0A-29-B1-CD-10-7D-5B-1E-99-4A-15-AB-77-AD-00-00-00-00
```
with the below exception
```
System.ArgumentOutOfRangeException: Index was out of range. Must be non-negative and less than the size of the collection. (Parameter 'startIndex')
  Module "System.ThrowHelper", in ThrowArgumentOutOfRangeException
  Module "System.BitConverter", in ToInt32
  Module "System.BitConverter", in ToUInt32
  Module "Registry.Cells.LkCellRecord", in .ctor
  Module "Registry.Other.HBinRecord", in Process
```

I don't see this on every hive I attempt to parse, just occasionally. I am loading the hive with the following options:
```csharp
LoadedHive = new RegistryHive(expandedPath)
                {
                    RecoverDeleted = false,
                    FlushRecordListsAfterParse = true
                };
LoadedHive.ParseHive();
```

I am parsing COMPONENTS hives - I haven't tried using any other hives. 

Is this something to worry about or can the error be safely ignored? It seems that I can still view the contents of the hive even after the error.

I can send you some hives that cause the issue if that would help - just let me know the best way to send them to you.

**Comment by EricZimmerman** [comment]
That is not a good error. That means data that should be there is not. How did you collect the hives? I only ever see that error when the hives are damaged. Make sure you collect the corresponding LOG files too

It's good you still get data but something would be missing from the subkeys or a keys values with a hard error. 

I assume you are using the latest nuget package?

**Comment by TeknoVenus** [comment]
I'm using the latest version - 1.1.0.5

I am actually writing code to detect corruptions in the COMPONENTS hive, so it's likely some of the hives I'm testing do have damage. Is it possible to determine which key/value caused the error?

I wonder if the issue relates to LOG files, as I'm just collecting and parsing the COMPONENTS hive files themselves and not the LOG files. That said, the hives were all properly unloaded from the registry before they were collected, so I wouldn't expect them to have any pending transactions.

**Comment by EricZimmerman** [comment]
My code would tell you if the registry hives were dirty which means you
need the logs, so if it's not complaining then the hives aren't dirty

You can always enable debug and/or Trace messages to get the exact offset
and location and everything else where you are running into your issue

**Comment by TeknoVenus** [comment]
Ah ok, didn't realise that. It's not complaining about dirty hives so that rules that out

With debug logging, it doesn't reveal much more useful to my eyes. I see the Trace logging is all commented out in the Registry code so I'll need to uncomment and rebuild locally to get some Trace level logs

```
2020-10-14 22:29:04.6589|INFO|ComponentsCollectorNET.Hive|Parsing loaded hive, please wait... 
2020-10-14 22:29:04.6684|DEBUG|D:\Stephen\Desktop\components_repaired|Header length is smaller than the size of the file. 
2020-10-14 22:29:05.2307|ERROR|D:\Stephen\Desktop\components_repaired|Hard error processing record with cell signature lk at Absolute Offset: 0x2632570 with raw data: E8-FF-FF-FF-6C-6B-7A-C3-B6-75-4B-97-F6-53-EE-49-42-95-59-47-91-C3-03-E9 System.ArgumentOutOfRangeException Index was out of range. Must be non-negative and less than the size of the collection. (Parameter 'startIndex') Void ThrowArgumentOutOfRangeException(System.ExceptionArgument, System.ExceptionResource)
2020-10-14 22:29:05.3817|DEBUG|D:\Stephen\Desktop\components_repaired|Found hbin with size 0 at absolute offset 0x2B30000 
2020-10-14 22:29:05.3908|DEBUG|D:\Stephen\Desktop\components_repaired|Initial processing complete. Building tree... 
2020-10-14 22:29:05.4022|DEBUG|D:\Stephen\Desktop\components_repaired|Found root node! Getting subkeys... 
2020-10-14 22:29:05.4022|DEBUG|D:\Stephen\Desktop\components_repaired|Created root node object. Getting subkeys. 
2020-10-14 22:29:06.4303|DEBUG|D:\Stephen\Desktop\components_repaired|Hive processing complete! 
2020-10-14 22:29:06.4357|DEBUG|D:\Stephen\Desktop\components_repaired|Flushing record lists... 
```

**Comment by EricZimmerman** [comment]
hmm, i dont recall why i would have commented out the trace lines since its just nlog, but i think it was because most of the output was overwhelming. i should probably revisit it

the error in the data itself is at 0x2632570

i would probably download my registry project and set up a unit test with the hive you are using. then you can walk thru that and see the exact location of the error, which key/value its for, etc.

thats the easiest way.

**Comment by TeknoVenus** [comment]
Have put a unit test together so I can debug this

It's failing to construct the LkCellRecord. Can't find much documentation on LkCellRecords, but I think this means it's not a specific key/value that's causing the issue here.

The raw byte array passed to the constructor is of length 0x18, which is too short as the rest of the constructor is expecting to read values at much higher indexes. Exception is therefore thrown here:
```csharp
//RootCellIndex
var num = BitConverter.ToUInt32(rawBytes, 0x20);
```

In `HBinRecord`, the recordSize is `0xffffffe8`, which overflows when converted to int to -24. Looking at your comments this is an expected situation though?
```csharp
// if we get a negative number here the record is allocated, but we cant read negative bytes, so get absolute value
```
However, since this is an LK record, the constructor is expecting a much longer array of bytes so I don't think this is an expected situation for a LK record.

Interestingly, if I open the hive in RegistryExplorer, I get a second error I don't get when parsing the hive myself. Assume this is because RegistryExplorer loads the hive with different options?

```
Message Date	Message Type	Message	Hive Path
2020-10-14 23:52:24	Info	Registry Explorer version 1.5.2.0 ready!	
2020-10-14 23:53:09	Warning	Soft error processing record with cell signature lk at Absolute Offset: 0x10865D0 with raw data: 10-00-00-00-6C-6B-6F-65-73-52-49-3D-01-00-0C-00	D:\Stephen\Desktop\components_github
2020-10-14 23:53:09	Error	Hard error processing record with cell signature lk at Absolute Offset: 0x2632570 with raw data: E8-FF-FF-FF-6C-6B-7A-C3-B6-75-4B-97-F6-53-EE-49-42-95-59-47-91-C3-03-E9	D:\Stephen\Desktop\components_github
```

I'll have to do some more investigation another day as run out of time today, but this is just some notes for now...

If you want to/have chance to have a look (no worries if not), I've put a copy of the hive here: https://stephenfoulds.com/components_github.zip.

Thanks for your help and the very quick responses - Registry is a great project!

**Comment by EricZimmerman** [comment]
so the issue here is that when that error happens, the keys and subkeys are not resolved,

soft error is when data recovery is being attempted, those are OK. you are seeing that error in RE because it will by default, try to recover deleted stuff.

i have quite a bit of registry stuff written up at my blog:

https://binaryforay.blogspot.com/2015/01/registry-hive-basics.html


all In use hbins, NKs, etc have a negative size, so that is expected. it means its in use.

i would be interested how many other places lk records show up. would be interesting to do a bunch of other hives that work to see if you get them and what they look like by comparison.

**Comment by TeknoVenus** [comment]
I've collected a lot of different COMPONENTS hives for testing, all from completely different computers running a mix of different versions of Windows 7 and newer.

I wrote a quick program that parses all those hives and see what errors occur. Most of the hives parsed fine, but quite a few had the same error parsing LK records. However, the hives that I tested report an LK count of 0, so I can't compare with a good hive to see what the LK record should look like. It seems that if the hive has an LK record, Registry always fails to parse it.

I don't think this is an issue with damaged hives, as I find it odd that all the hives would be damaged in the same way.

Any thoughts?

Here's the output from my program:

```
Reigstry test program
INFO | ================== Loading hive D:\Stephen\Desktop\ComponentsHives\001_ComponentsHive ==================
INFO | Found 7,563 hbin records. Total size of seen hbin records: 0x1F7F000, Header hive size: 0x1F6E000
INFO | Found 258,523 Cell records (nk: 62,922, vk: 195,593, sk: 8, lk: 0)
INFO | 0 hard parsing errors
INFO | 0 soft parsing errors

INFO | ================== Loading hive D:\Stephen\Desktop\ComponentsHives\002_ComponentsHive ==================
INFO | Found 13,258 hbin records. Total size of seen hbin records: 0x37CA000, Header hive size: 0x37CA000
INFO | Found 337,004 Cell records (nk: 81,748, vk: 255,249, sk: 7, lk: 0)
INFO | 0 hard parsing errors
INFO | 0 soft parsing errors

INFO | ================== Loading hive D:\Stephen\Desktop\ComponentsHives\003_ComponentsHive ==================
INFO | Found 5,577 hbin records. Total size of seen hbin records: 0x16BF000, Header hive size: 0x16A9000
INFO | Found 180,625 Cell records (nk: 57,339, vk: 123,280, sk: 6, lk: 0)
INFO | 0 hard parsing errors
INFO | 0 soft parsing errors

INFO | ================== Loading hive D:\Stephen\Desktop\ComponentsHives\004_ComponentsHive ==================
WARN | Sequence numbers do not match! Hive is dirty and the transaction logs should be reviewed for relevant data!
ERROR | Hard error processing record with cell signature lk at Absolute Offset: 0x1392490 with raw data: D8-FF-FF-FF-6C-6B-40-16-F2-88-84-09-C3-8A-8D-13-CC-EB-51-3C-3C-49-0A-29-B1-CD-10-7D-5B-1E-99-4A-15-AB-77-AD-00-00-00-00
ERROR | Hard error processing record with cell signature lk at Absolute Offset: 0x1C91870 with raw data: D8-FF-FF-FF-6C-6B-72-48-75-6E-EE-CF-5C-45-D6-64-BD-80-63-D4-ED-1F-56-C2-BE-66-8E-37-6E-C7-7E-6F-04-12-8F-03-00-00-00-00
INFO | Found 12,905 hbin records. Total size of seen hbin records: 0x3697000, Header hive size: 0x3697000
INFO | Found 449,890 Cell records (nk: 142,744, vk: 307,138, sk: 8, lk: 0)
INFO | 2 hard parsing errors
INFO | 0 soft parsing errors

INFO | ================== Loading hive D:\Stephen\Desktop\ComponentsHives\005_ComponentsHive ==================
ERROR | Hard error processing record with cell signature lk at Absolute Offset: 0x1C41758 with raw data: D8-FF-FF-FF-6C-6B-4F-CE-81-94-10-30-34-40-EA-57-13-55-6A-37-4F-39-27-C7-31-DF-16-54-F2-E5-DB-5F-81-C6-9C-5F-47-1D-A7-9D
INFO | Found 14,415 hbin records. Total size of seen hbin records: 0x3A9B000, Header hive size: 0x3A9B000
INFO | Found 493,823 Cell records (nk: 116,639, vk: 377,177, sk: 7, lk: 0)
INFO | 1 hard parsing errors
INFO | 0 soft parsing errors

INFO | ================== Loading hive D:\Stephen\Desktop\ComponentsHives\006_ComponentsHive ==================
INFO | Found 7,345 hbin records. Total size of seen hbin records: 0x1D10000, Header hive size: 0x1D10000
INFO | Found 250,305 Cell records (nk: 69,788, vk: 180,512, sk: 5, lk: 0)
INFO | 0 hard parsing errors
INFO | 0 soft parsing errors

INFO | ================== Loading hive D:\Stephen\Desktop\ComponentsHives\007_ComponentsHive ==================
INFO | Found 11,874 hbin records. Total size of seen hbin records: 0x3008000, Header hive size: 0x3008000
INFO | Found 395,280 Cell records (nk: 102,040, vk: 293,235, sk: 5, lk: 0)
INFO | 0 hard parsing errors
INFO | 0 soft parsing errors

INFO | ================== Loading hive D:\Stephen\Desktop\ComponentsHives\008_ComponentsHive ==================
ERROR | Hard error processing record with cell signature lk at Absolute Offset: 0x2915E00 with raw data: D8-FF-FF-FF-6C-6B-36-51-B5-14-BA-2E-24-65-8A-AD-F0-66-C0-F4-69-E0-56-73-66-04-35-91-C4-6C-21-D1-56-B4-EA-45-33-36-36-2D
ERROR | Hard error processing record with cell signature lk at Absolute Offset: 0x9ACA6B0 with raw data: E8-FF-FF-FF-6C-6B-7E-6A-F6-17-F9-48-07-0F-6C-C1-3A-25-9F-CF-71-AD-06-71
ERROR | Hard error processing record with cell signature lk at Absolute Offset: 0xAABB8F8 with raw data: D8-FF-FF-FF-6C-6B-42-31-06-2A-37-EF-30-AA-FD-84-61-33-1A-45-8C-77-3D-AE-DB-95-E7-C6-41-26-19-7A-D2-15-40-E0-00-00-00-00
ERROR | Hard error processing record with cell signature lk at Absolute Offset: 0xAACC750 with raw data: E8-FF-FF-FF-6C-6B-F0-EE-96-35-0A-FA-4F-54-8C-18-DF-E1-0A-8F-9B-F1-61-CB
INFO | Found 47,607 hbin records. Total size of seen hbin records: 0xBB7C000, Header hive size: 0xBB7C000
INFO | Found 1,711,875 Cell records (nk: 481,549, vk: 1,230,319, sk: 7, lk: 0)
INFO | 4 hard parsing errors
INFO | 0 soft parsing errors

INFO | ================== Loading hive D:\Stephen\Desktop\ComponentsHives\009_ComponentsHive ==================
INFO | Found 11,442 hbin records. Total size of seen hbin records: 0x2D7D000, Header hive size: 0x2A93000
INFO | Found 310,971 Cell records (nk: 88,714, vk: 222,248, sk: 9, lk: 0)
INFO | 0 hard parsing errors
INFO | 0 soft parsing errors

INFO | ================== Loading hive D:\Stephen\Desktop\ComponentsHives\010_ComponentsHive ==================
INFO | Found 11,669 hbin records. Total size of seen hbin records: 0x2E2A000, Header hive size: 0x2E2A000
INFO | Found 252,173 Cell records (nk: 72,704, vk: 179,464, sk: 5, lk: 0)
INFO | 0 hard parsing errors
INFO | 0 soft parsing errors

INFO | ================== Loading hive D:\Stephen\Desktop\ComponentsHives\011_ComponentsHive ==================
WARN | Sequence numbers do not match! Hive is dirty and the transaction logs should be reviewed for relevant data!
INFO | Found 21,400 hbin records. Total size of seen hbin records: 0x594A000, Header hive size: 0x4577000
INFO | Found 737,918 Cell records (nk: 191,861, vk: 546,051, sk: 6, lk: 0)
INFO | 0 hard parsing errors
INFO | 0 soft parsing errors

INFO | ================== Loading hive D:\Stephen\Desktop\ComponentsHives\012_ComponentsHive ==================
INFO | Found 12,856 hbin records. Total size of seen hbin records: 0x32D5000, Header hive size: 0x32D5000
INFO | Found 330,514 Cell records (nk: 91,108, vk: 239,401, sk: 5, lk: 0)
INFO | 0 hard parsing errors
INFO | 0 soft parsing errors

INFO | ================== Loading hive D:\Stephen\Desktop\ComponentsHives\013_ComponentsHive ==================
WARN | Sequence numbers do not match! Hive is dirty and the transaction logs should be reviewed for relevant data!
INFO | Found 18,290 hbin records. Total size of seen hbin records: 0x4D2D000, Header hive size: 0x4D2D000
INFO | Found 409,766 Cell records (nk: 96,703, vk: 313,054, sk: 9, lk: 0)
INFO | 0 hard parsing errors
INFO | 0 soft parsing errors

INFO | ================== Loading hive D:\Stephen\Desktop\ComponentsHives\014_ComponentsHive ==================
WARN | Sequence numbers do not match! Hive is dirty and the transaction logs should be reviewed for relevant data!
ERROR | Hard error processing record with cell signature lk at Absolute Offset: 0xD8C3D8 with raw data: D8-FF-FF-FF-6C-6B-40-16-F2-88-84-09-C3-8A-8D-13-CC-EB-51-3C-3C-49-0A-29-B1-CD-10-7D-5B-1E-99-4A-15-AB-77-AD-00-00-00-00
INFO | Found 9,815 hbin records. Total size of seen hbin records: 0x293F000, Header hive size: 0x2917000
INFO | Found 337,393 Cell records (nk: 84,961, vk: 252,424, sk: 8, lk: 0)
INFO | 1 hard parsing errors
INFO | 0 soft parsing errors

INFO | ================== Loading hive D:\Stephen\Desktop\ComponentsHives\015_ComponentsHive ==================
ERROR | Hard error processing record with cell signature lk at Absolute Offset: 0x17819D0 with raw data: D8-FF-FF-FF-6C-6B-40-16-F2-88-84-09-C3-8A-8D-13-CC-EB-51-3C-3C-49-0A-29-B1-CD-10-7D-5B-1E-99-4A-15-AB-77-AD-00-00-00-00
INFO | Found 21,003 hbin records. Total size of seen hbin records: 0x5827000, Header hive size: 0x5827000
INFO | Found 478,305 Cell records (nk: 118,835, vk: 359,462, sk: 8, lk: 0)
INFO | 1 hard parsing errors
INFO | 0 soft parsing errors

INFO | ================== Loading hive D:\Stephen\Desktop\ComponentsHives\016_ComponentsHive ==================
WARN | Sequence numbers do not match! Hive is dirty and the transaction logs should be reviewed for relevant data!
ERROR | Hard error processing record with cell signature lk at Absolute Offset: 0x1199D60 with raw data: D8-FF-FF-FF-6C-6B-40-16-F2-88-84-09-C3-8A-8D-13-CC-EB-51-3C-3C-49-0A-29-B1-CD-10-7D-5B-1E-99-4A-15-AB-77-AD-00-00-00-00
WARN | hbin header incorrect at absolute offset 0x275E000!!! Percent done: 99.66%
INFO | Found 9,398 hbin records. Total size of seen hbin records: 0x275D000, Header hive size: 0x275B000
INFO | Found 319,229 Cell records (nk: 81,445, vk: 237,777, sk: 7, lk: 0)
INFO | 1 hard parsing errors
INFO | 0 soft parsing errors

INFO | ================== Loading hive D:\Stephen\Desktop\ComponentsHives\017_ComponentsHive ==================
ERROR | Hard error processing record with cell signature lk at Absolute Offset: 0x2632570 with raw data: E8-FF-FF-FF-6C-6B-7A-C3-B6-75-4B-97-F6-53-EE-49-42-95-59-47-91-C3-03-E9
INFO | Found 10,860 hbin records. Total size of seen hbin records: 0x2B2F000, Header hive size: 0x2B2F000
INFO | Found 302,103 Cell records (nk: 84,739, vk: 217,357, sk: 7, lk: 0)
INFO | 1 hard parsing errors
INFO | 0 soft parsing errors

INFO | ================== Loading hive D:\Stephen\Desktop\ComponentsHives\018_ComponentsHive ==================
INFO | Found 8,730 hbin records. Total size of seen hbin records: 0x271B000, Header hive size: 0x271B000
INFO | Found 297,488 Cell records (nk: 86,505, vk: 210,975, sk: 8, lk: 0)
INFO | 0 hard parsing errors
INFO | 0 soft parsing errors

INFO | ================== Loading hive D:\Stephen\Desktop\ComponentsHives\019_ComponentsHive ==================
WARN | Sequence numbers do not match! Hive is dirty and the transaction logs should be reviewed for relevant data!
WARN | hbin header incorrect at absolute offset 0x1EE2000!!! Percent done: 99.62%
INFO | Found 7,390 hbin records. Total size of seen hbin records: 0x1EE1000, Header hive size: 0x1EE1000
INFO | Found 254,831 Cell records (nk: 62,177, vk: 192,646, sk: 8, lk: 0)
INFO | 0 hard parsing errors
INFO | 0 soft parsing errors

INFO | ================== Loading hive D:\Stephen\Desktop\ComponentsHives\020_ComponentsHive ==================
INFO | Found 12,051 hbin records. Total size of seen hbin records: 0x32AC000, Header hive size: 0x32AC000
INFO | Found 427,627 Cell records (nk: 100,655, vk: 326,963, sk: 9, lk: 0)
INFO | 0 hard parsing errors
INFO | 0 soft parsing errors

INFO | ================== Loading hive D:\Stephen\Desktop\ComponentsHives\021_ComponentsHive ==================
INFO | Found 6,629 hbin records. Total size of seen hbin records: 0x1BC4000, Header hive size: 0x1BC3000
INFO | Found 215,182 Cell records (nk: 66,393, vk: 148,783, sk: 6, lk: 0)
INFO | 0 hard parsing errors
INFO | 0 soft parsing errors

INFO | ================== Loading hive D:\Stephen\Desktop\ComponentsHives\022_ComponentsHive ==================
INFO | Found 6,700 hbin records. Total size of seen hbin records: 0x1B14000, Header hive size: 0x1B14000
INFO | Found 222,921 Cell records (nk: 62,568, vk: 160,347, sk: 6, lk: 0)
INFO | 0 hard parsing errors
INFO | 0 soft parsing errors

INFO | ================== Loading hive D:\Stephen\Desktop\ComponentsHives\023_ComponentsHive ==================
INFO | Found 6,661 hbin records. Total size of seen hbin records: 0x1AED000, Header hive size: 0x1AED000
INFO | Found 221,178 Cell records (nk: 62,356, vk: 158,816, sk: 6, lk: 0)
INFO | 0 hard parsing errors
INFO | 0 soft parsing errors

INFO | ================== Loading hive D:\Stephen\Desktop\ComponentsHives\024_ComponentsHive ==================
INFO | Found 7,147 hbin records. Total size of seen hbin records: 0x1C3F000, Header hive size: 0x1C37000
INFO | Found 212,056 Cell records (nk: 50,935, vk: 161,116, sk: 5, lk: 0)
INFO | 0 hard parsing errors
INFO | 0 soft parsing errors

INFO | ================== Loading hive D:\Stephen\Desktop\ComponentsHives\025_ComponentsHive ==================
ERROR | Hard error processing record with cell signature lk at Absolute Offset: 0x1393D70 with raw data: D8-FF-FF-FF-6C-6B-4F-CE-81-94-10-30-34-40-EA-57-13-55-6A-37-4F-39-27-C7-31-DF-16-54-F2-E5-DB-5F-81-C6-9C-5F-76-6B-03-00
INFO | Found 15,473 hbin records. Total size of seen hbin records: 0x3D51000, Header hive size: 0x3D49000
INFO | Found 345,995 Cell records (nk: 68,145, vk: 277,842, sk: 8, lk: 0)
INFO | 1 hard parsing errors
INFO | 0 soft parsing errors

INFO | ================== Loading hive D:\Stephen\Desktop\ComponentsHives\026_ComponentsHive ==================
INFO | Found 8,508 hbin records. Total size of seen hbin records: 0x219D000, Header hive size: 0x219D000
INFO | Found 207,760 Cell records (nk: 56,209, vk: 151,547, sk: 4, lk: 0)
INFO | 0 hard parsing errors
INFO | 0 soft parsing errors

INFO | ================== Loading hive D:\Stephen\Desktop\ComponentsHives\027_ComponentsHive ==================
INFO | Found 15,394 hbin records. Total size of seen hbin records: 0x3CFE000, Header hive size: 0x3CFE000
INFO | Found 501,353 Cell records (nk: 117,667, vk: 383,681, sk: 5, lk: 0)
INFO | 0 hard parsing errors
INFO | 0 soft parsing errors

INFO | ================== Loading hive D:\Stephen\Desktop\ComponentsHives\028_ComponentsHive ==================
INFO | Found 10,640 hbin records. Total size of seen hbin records: 0x2A05000, Header hive size: 0x2A05000
INFO | Found 221,432 Cell records (nk: 65,064, vk: 156,363, sk: 5, lk: 0)
INFO | 0 hard parsing errors
INFO | 0 soft parsing errors

INFO | ================== Loading hive D:\Stephen\Desktop\ComponentsHives\029_ComponentsHive ==================
INFO | Found 10,642 hbin records. Total size of seen hbin records: 0x2A14000, Header hive size: 0x2A14000
INFO | Found 212,897 Cell records (nk: 66,872, vk: 146,019, sk: 6, lk: 0)
INFO | 0 hard parsing errors
INFO | 0 soft parsing errors

INFO | ================== Loading hive D:\Stephen\Desktop\ComponentsHives\030_ComponentsHive ==================
WARN | Sequence numbers do not match! Hive is dirty and the transaction logs should be reviewed for relevant data!
INFO | Found 10,820 hbin records. Total size of seen hbin records: 0x2D3F000, Header hive size: 0x2D10000
INFO | Found 381,531 Cell records (nk: 90,146, vk: 291,378, sk: 7, lk: 0)
INFO | 0 hard parsing errors
INFO | 0 soft parsing errors
```

**Comment by msuhanov** [comment]
> If you want to/have chance to have a look (no worries if not), I've put a copy of the hive here: https://stephenfoulds.com/components_github.zip.

> 2020-10-14 23:53:09	Error	Hard error processing record with cell signature lk at Absolute Offset: 0x2632570 with raw data: E8-FF-FF-FF-**6C-6B-7A-C3-B6-75-4B-97-F6-53-EE-49-42-95-59-47-91-C3-03-E9**	D:\Stephen\Desktop\components_github

In the registry hive you posted, the bytes mentioned belong to value data, not to a subkeys list.

```
Key path: DerivedData\VersionedIndex\6.1.7601.18766 (win7sp1_gdr.150217-1551)\ComponentFamilies\amd64_microsoft-windows-i..ersandsecurityzones_31bf3856ad364e35_none_de3177239ae782e5\v!10.2.9200.17267
Last written timestamp (UTC): 2020-09-28 02:49:35.347802
Access bits: 0
Owner SID: S-1-5-18

<...>

Value name: S1H
Value type: REG_BINARY
Data size: 20
Data (hexdump):
00000000  6C 6B 7A C3 B6 75 4B 97-F6 53 EE 49 42 95 59 47  lkz..uK..S.IB.YG
00000010  91 C3 03 E9                                      ....

<...>
```

For some reason, the parser tries to interpret these bytes as a subkeys list. This should never be done without going to a key node first.

The bug is here: https://github.com/EricZimmerman/Registry/blob/342eb4757ece31bbda184196e359e06d4ec97cec/Registry/RegistryHive.cs#L978

**Comment by EricZimmerman** [comment]
@msuhanov thanks for taking a look. i will get that fixed asap!

**Comment by EricZimmerman** [comment]
i just pushed nuget 1.1.1 which should fix  this.

that hive parsed without the 2 parsing errors. i had to add a check for the length when i saw lk signature. thats the first hive i have seen where i ran into that issue. thank you for reporting this!

please let me know how 1.1.1 works for you and if i ned to reopen this i will.

**Comment by TeknoVenus** [comment]
Just tested this with my collection of hives and all of them loaded without any errors, so this seems to be fixed in 1.1.1

Thanks for investigating and fixing!
